pythonforandroid/build.py

```diff
@@ -709,25 +703,157 @@ def run_pymodules_install(ctx, modules):
         env = copy.copy(base_env)
         env.update(recipe_env)
 
-        info('Installing Python modules with pip')
-        info('IF THIS FAILS, THE MODULES MAY NEED A RECIPE. '
-             'A reason for this is often modules compiling '
-             'native code that is unaware of Android cross-compilation '
-             'and does not work without additional '
-             'changes / workarounds.')
-
         # Make sure our build package dir is available, and the virtualenv
-        # site packages come FIRST (for the proper pip version):
+        # site packages come FIRST (so the proper pip version is used):
         env["PYTHONPATH"] += ":" + ctx.get_site_packages_dir()
         env["PYTHONPATH"] = os.path.abspath(join(
             ctx.build_dir, "venv", "lib",
             "python" + ctx.python_recipe.major_minor_version_string,
             "site-packages")) + ":" + env["PYTHONPATH"]
-        shprint(sh.bash, '-c', (
-            "source venv/bin/activate && " +
-            "pip install -v --target '{0}' --no-deps -r requirements.txt"
-        ).format(ctx.get_site_packages_dir().replace("'", "'\"'\"'")),
-                _env=copy.copy(env))
+
+        # Install the manually specified requirements first:
+        if not modules:
+            info('There are no Python modules to install, skipping')
+        else:
+            info('Creating a requirements.txt file for the Python modules')
+            with open('requirements.txt', 'w') as fileh:
+                for module in modules:
+                    key = 'VERSION_' + module
+                    if key in environ:
+                        line = '{}=={}\n'.format(module, environ[key])
```

Comment:
Out of context since you just moved these lines around, but I was curious why this is written that way :thinking:

Reply:
I don't think there's a good reason. It is stored in the env in `toolchain.py` and dug out of the env here, it should be a member on the `Context()` instead. Pretty sure this is just a spontaneous idea gone meh which is worth refactoring, but this pull request is already big enough